core/src/multi_proof_verification.rs

```diff
@@ -0,0 +1,636 @@
+use crate::{
+    multi_proof::{MultiProof, SubPathProof},
+    proof::{
+        hash_path, shared_bits, PathProofTerminal, PathProofVerificationError, VerifiedPathProof,
+    },
+    trie::{InternalData, KeyPath, LeafData, Node, NodeHasher, NodeHasherExt, TERMINATOR},
+};
+use alloc::collections::BTreeMap;
+use bitvec::{order::Msb0, vec::BitVec, view::BitView};
+
+/// A verifiable proof of multiple path through the trie.
+#[derive(Debug, Clone)]
+pub struct VerifiableMultiProof<H: NodeHasher> {
+    nodes: BTreeMap<BitVec<u8, Msb0>, Node>,
```

Comment:
these BTreeMaps seem unnecessary. AFAIK all tree algorithms that construct nodes only need a stack. usually that's a left-stack, and I believe that still holds in this bisection example, because the left bisection is always created first.